File: coding/beautiful_arrangement.c
#include<stdio.h>
#include<stdlib.h>

void helper(int, int*, int, int *);

int countArrangement(int N) {
    int *nums = (int*) malloc(sizeof(int)*(N+1));
    int i, count = 0, cur_index = 1;
    for(i = 1; i <= N; ++i) {
        nums[i] = i;
    }
    
    helper(N, nums, cur_index, &count);
    return count;
}

void swap(int *a, int *b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

void helper(int N, int *nums, int index, int *count) {
    int i;
    if (index == N+1) {
        *count += 1;
    }else {
        for(i = index; i <= N; ++i) {
            swap(&nums[i], &nums[index]);
            if ( (index % nums[index] == 0) || (nums[index] % (index) == 0) ) {
                helper(N, nums, index+1, count);
            }
            swap(&nums[i], &nums[index]);
        }
    }
}

int main()
{
	printf("%d\n",countArrangement(15));
	return 0;
}

File: coding/beautiful_arrangement.py
class Solution:
    def countArrangement(self, N):
        """
        :type N: int
        :rtype: int
        """
        nums = [ x for x in range(1, N+1) ]
        return self.count_helper(nums, 0, 0)
    
    def count_helper(self, nums, cur_index=0, count=0):
        if cur_index == len(nums):
            count += 1
        else:
            for i in range(cur_index, len(nums)):
                nums[cur_index], nums[i] = nums[i], nums[cur_index]
                if (nums[cur_index] % (cur_index+1) == 0) or ((cur_index+1) % nums[cur_index] == 0):
                    count = self.count_helper(nums, cur_index+1, count)
                nums[cur_index], nums[i] = nums[i], nums[cur_index]
        return count
    

s = Solution()
print(s.countArrangement(15))   
